cranelift/isle/isle/src/trie_again.rs

```diff
@@ -0,0 +1,594 @@
+            }
+        }
+    }
+
+    fn field_bindings(
+        &mut self,
+        binding: BindingId,
+        fields: TupleIndex,
+        variant: sema::VariantId,
+        deferred_constraints: &mut Vec<(BindingId, Constraint)>,
+    ) -> Box<[BindingId]> {
+        (0..fields.0)
+            .map(TupleIndex)
+            .map(move |field| {
+                let binding = self.dedup_binding(Binding::Variant {
+                    source: binding,
+                    variant,
+                    field,
+                });
+                // We've just added an equality constraint to a binding site that may not have had
+                // one already. If that binding site already had a concrete constraint, then we need
+                // to "recursively" propagate that constraint through the new equivalence class too.
+                if let Some(constraint) = self.current_rule.get_constraint(binding) {
+                    deferred_constraints.push((binding, constraint));
+                }
+                binding
+            })
+            .collect()
+    }
+
+    fn dedup_binding(&mut self, binding: Binding) -> BindingId {
+        if let Some(binding) = self.binding_map.get(&binding) {
+            *binding
+        } else {
+            let id = BindingId(self.rules.bindings.len().try_into().unwrap());
+            self.rules.bindings.push(binding.clone());
+            self.binding_map.insert(binding, id);
+            id
+        }
+    }
+
+    fn dedup_expr(&mut self, expr: Expr) -> ExprId {
+        if let Some(expr) = self.expr_map.get(&expr) {
+            *expr
+        } else {
+            let id = ExprId(self.rules.exprs.len().try_into().unwrap());
+            self.rules.exprs.push(expr.clone());
+            self.expr_map.insert(expr, id);
+            id
+        }
+    }
+
+    fn set_constraint(&mut self, input: Binding, constraint: Constraint) -> BindingId {
+        let input = self.dedup_binding(input);
+        if let Err(e) = self.current_rule.set_constraint(input, constraint) {
+            self.unreachable.push(e);
+        } else {
+            self.set_constraint_on_class(input, constraint);
+        }
+        input
+    }
+}
+
+impl sema::PatternVisitor for RuleSetBuilder {
+    type PatternId = Binding;
+
+    fn add_match_equal(&mut self, a: Binding, b: Binding, _ty: sema::TypeId) {
+        let a = self.dedup_binding(a);
+        let b = self.dedup_binding(b);
+        // If both bindings represent the same binding site, they're implicitly equal.
+        if a != b {
+            self.current_rule.equals.merge(a, b);
+            // If either binding had a constraint, copy that constraint to the whole equivalence
+            // class (including the other binding).
+            if let Some(constraint) = self.current_rule.get_constraint(a) {
+                self.set_constraint_on_class(a, constraint);
+            } else if let Some(constraint) = self.current_rule.get_constraint(b) {
```

Comment:
Add a comment here describing why safe to only check for constraint on `b` if we didn't invoke `set_constraint_on_class(a, ...)` above -- because the above will otherwise see any constraints on `b` because it's part of the same equivalence class now (due to merge above).